File: campaign_manager/gos/serializers/monster.py
from rest_framework import serializers

from gos.models.monster import Monster
from gos.serializers.npc_class import NPCClassSerializer
from gos.serializers.spell_slot import SpellSlotSerializer


class MonsterDetailSerializer(serializers.ModelSerializer):
    spell_slots = SpellSlotSerializer(many=True, required=False)
    npc_class = NPCClassSerializer(required=False)

    class Meta:
        model = Monster
        fields = [
            'id',
            'name',
            'npc_class',
            'current_hp',
            'spell_slots'
        ]

class MonsterListSerializer(serializers.ModelSerializer):
    class Meta:
        model = Monster
        fields = ['id', 'name', 'current_hp']

File: src/app/class/monster/monster.ts
import { NpcClass } from '../npc-class/npc-class';
import { SpellSlot } from '../spell-slot/spell-slot';

export class Monster {
    id: number;
    name: string;
    npcClass: NpcClass;
    currentHp: number;
    spellSlots: SpellSlot[];
    initiative: number;

    constructor(props: any) {
        if (!props) return;

        this.id = props.id;
        this.name = props.name;
        this.npcClass = new NpcClass(props.npc_class);
        this.currentHp = props.current_hp;
        this.initiative = props.initiative;
        if (props.spell_slots) {
            this.spellSlots = props.spell_slots.map((slot: any) => new SpellSlot(slot));
        }
    }
}
